### PR TITLE
hasStrings Mac

### PR DESCRIPTION
**Pull request by justinmc**
## Description

Adding the hasStrings method to Mac, similar to https://github.com/flutter/engine/pull/19859.

## Related Issues

Depends on https://github.com/flutter/buildroot/pull/399
https://github.com/flutter/engine/pull/19859
https://github.com/flutter/flutter/issues/60145

## Tests

I added the following tests:

  * hasStrings returns true after setting a string to the clipboard.
  * hasStrings returns false after clearing the clipboard.

### CODE REVIEW

**Comment by justinmc** [comment]
What's the status of unit testing on Mac? 
 I see one unit test, and it doesn't seem to be run by run_tests.py: https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterEngineUnittests.mm

**Comment by justinmc** [comment]
Whoops, actually I forgot to remove the line referencing it when I was messing around with trying to figure out the Mac unit tests.

Do we actually have functioning unit tests on Mac?  If so, how do I run them?

**Comment by gaaclarke** [comment]
> Do we actually have functioning unit tests on Mac? If so, how do I run them?

Never looked into it.  @stuartmorgan would be the person to ask.

**Comment by justinmc** [comment]
Thanks for the pointer! I couldn't figure this out and didn't know who to ask.

**Comment by stuartmorgan** [comment]
The current state of treating for macOS is not good, certainly. @chinmaygarde added that test so could answer questions about it, or I can look next week when I'm not OOO

**Comment by justinmc** [comment]
@stuartmorgan Thanks for the pointer, I was able to get the tests running with help from Chinmay.  I will attempt to add the Mac unit tests into run_tests.py in this PR.

For posterity, the way to directly run the Mac unit tests is to compile for Mac, then run:

`./out/host_debug_unopt/flutter_desktop_darwin_unittests`

**Comment by justinmc** [comment]
@gaaclarke Ready for re-review.  I was able to add a test and get the Mac unit tests running as a part of run_tests.py.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:13]: Why is this called MacOSTestTest? The name should describe what the test is covering.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:43]: I suspect the unit tests on the bots are running in a bootstrap context that doesn't have access to the pasteboard server. If that's the case, perhaps the tests should inject a mock `NSPasteboard`.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:13]: Thanks, forgot to update that.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:43]: Any idea how to include OCMock in the Mac tests?  I tried including it in this test file and adding it to Build.gn in various places, but I keep getting compiler errors.

Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:43]: The same way it's done for iOS? Without knowing what you tried and what exactly didn't work, it's hard to advise further.

**Comment by justinmc** [comment]
I'm still getting errors, but this is my attempt at getting OCMock to work in the Mac tests so far: https://github.com/flutter/engine/pull/20531/commits/6177e74e218c32c7fc98264a8281b849faa67e4e

This led me to our [BUILD.gn](https://github.com/flutter/buildroot/blob/master/build/secondary/third_party/ocmock/BUILD.gn) file for OCMock in the buildroot repository, which has some iOS stuff in it that I'll probably need to change (if I'm on the right track).  It asserts `is_ios` and also sets the minimum iOS simulator version. If I (temporarily) remove those lines, I can get it to start building before it crashes with:

```
flutter_desktop_darwin_unittests[10613:107792] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '** Expected method not present; the method getArgumentAtIndexAsObject: is not implemented by NSInvocation. If you see this exception it is likely that you are using the static library version of OCMock and your project is not configured correctly to load categories from static libraries. Did you forget to add the -ObjC linker flag?'
```

@gaaclarke Please let me know if you have any advice!

**Comment by gaaclarke** [comment]
@justinmc I was going to suggest something, but read the warning closer and it's already suggesting it haha.  Looks like you may be missing the "-ObjC" flag.

I'd get rid of the "assert ios" line, split out the code into a source_set, then make a new target which is "ocmock_mac" which depends on it, remove the ios version flag from there and try adding the "-ObjC" flag to it.  The iOS build must be getting the "-ObjC" from somewhere else unfortunately.

There is a weird bug in the objc compiler where you can't load objc classes.  They worked around it years ago by adding the "-ObjC" flag and never seemed to just make it work without the flag.

**Comment by justinmc** [comment]
I was able to get the test working with mocking, thanks again @gaaclarke.

This now depends on https://github.com/flutter/buildroot/pull/399, so once that is merged I'll update this branch and make sure the tests pass.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (gaaclarke) [inline shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:58]: This is non idiomatic objective-c, you don't use "getX" in objective c, you use properties, ex 
```objc 
@property(nonatomic, readonly, nonnull) NSPasteBoard* _pasteboard;
```

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (gaaclarke) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:547]: nit: It's a bit odd having a method called `clipboardHasStrings` return something other than a bool.  It is probably better to return bool and wrap it into a dictionary at the call site, imo.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (gaaclarke) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:28]: This code would be a bit easier to follow if you inlined all these blocks instead of declaring them then passing them as variables then using the variables.

**Review by gaaclarke — COMMENTED** [review]
Substance of the PR looks great.  Thanks for pushing on the testing.  I just have a few cleanup suggestions.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:58]: Why are you adding new public API to FVC to implement an internal channel? Unless there's a specific need for this to be public, it should be in the private header.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:56]: Why are you making this public?

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:180]: Why are you referencing UIKit documentation in the macOS embedding?

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:7]: This is the primary header for the test, so should be first in its own block.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:85]: This test is doing a lot of different things; please make individual tests that are more targeted. You're already using a fake pasteboard, so there's no need to do the entire cycle in a single test.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:56]: Good catch, I thought this was necessary to call it from the test, but it seems to work without it. I'll remove it.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:58]: Thanks for the tip!  This way seems cleaner and is just as mockable.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h]: I now have this as a property like the [comment](https://github.com/flutter/engine/pull/20531/files#r483133450) above recommends.  Let me know if how I did the property and getter doesn't look right. I just need to be able to mock it in the tests.

Excuse my lack of Objective C knowledge!

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:180]: There seems to be no equivalent hasStrings method for Mac.  I'll add some details in the comment.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:548]: That's a good call, the docs I wrote made this even more confusing actually.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:66]: Good call, I split it into two and it's a lot cleaner.

**Comment by justinmc** [comment]
@gaaclarke @stuartmorgan Ready for re-review.  Thank you guys for all the help on this!

**Comment by justinmc** [comment]
Anyone recognize these test failures?  They seem to be related to infrastructure, but merging in master didn't change anything.

```
[ERROR:flutter/runtime/dart_vm_data.cc(18)] VM snapshot invalid and could not be inferred from settings.
```

**Comment by stuartmorgan** [comment]
> Anyone recognize these test failures? They seem to be related to infrastructure, but merging in master didn't change anything.

Why do you think they are infrastructure-related? Does the new test binary you added to the script run locally for you in profile (or release) mode? Given that it wasn't previously being run on bots it seems much more likely to me that the test isn't properly set up to work correctly in AOT mode.

**Comment by justinmc** [comment]
Ah you're right, thanks!  Running in profile mode fails with the same error locally.

If you've got any idea how to get this working let me know.  The original FlutterEngineUnittests.mm also fails in profile mode.

**Comment by stuartmorgan** [comment]
> If you've got any idea how to get this working let me know. The original FlutterEngineUnittests.mm also fails in profile mode.

I'm not familiar with `GetFixturesPath()` and the associated scaffolding, but from a quick look I don't see any evidence that it's even been used with the iOS embedding either, so it may need non-trivial retooling to create AOT snapshots in the format macOS and iOS expect them; I'd have to dig deeper into it than I have time for at the moment to be sure. I'm also not sure that's a good use of effort at this point since for the desktop embedding tests in general I'd like to move toward stubbing out at the embedder.h layer (although we'll probably want some more full-stack tests for which GetFixturesPath() could still be useful).

At this point I think the best option for this PR would be to revert the addition of the test binary to `run_tests.py`. Needing to get it running on CI is a pre-existing issue, and you've added tests (and run them locally) covering your changes that are ready to come online when they are made to work on CI, so the testing situation has still improved incrementally as a result of the PR even if it's not at the ideal outcome yet.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:45]: This still needs to be moved to the private header.

Also, a property name should not start with `_`; the property and the ivar are distinct.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:180]: I guess I didn't explain my earlier question very well: what information is this adding that helps understand the method? The description in the comment seems clear and unambiguous; following that link doesn't tell me anything new AFAICT, so why have it?

It also seems odd to document it in terms of another platform's implementation, given than a) we have a lot of platforms, so referring to one of them is pretty arbitrary, b) that implementation is not relevant here since there's no such method on macOS, and c) that anyone who wants to know how it's done on iOS can easily just go look at the iOS source.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:531]: Prefer accessing properties with dot syntax: `self.pasteboard`.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController.mm:550]: Looking at this more closely this time: going through a method whose input and output are explicitly designed for the Flutter method channel contract seems like a very convoluted way to check this. Why not just `return [self.pasteboard stringForType:NSPasteboardTypeString]].length > 0;`?

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:43]: Your test controls the fake clipboard; there is no reason to involve `Clipboard.setData` in the test of Clipboard.hasStrings.

It would be good to have a test of `setData`, but that should be a separate test.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:72]: Same.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:45]: Where exactly should this be moved to?  Sorry for my lack of Objective C knowledge.  I tried moving this inside the .mm file, but then I couldn't mock it from the test.

👍  About the underscore.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewController.mm]: Sounds good, removed.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewController.mm]: Good point, done.

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm]: Done, now I just set up the mock with some given pasteboard string instead of letting the test call setString.

**Comment by justinmc** [comment]
@stuartmorgan That sounds good to me, I've reverted the change to run_tests.py.  I'm happy we were able to improve the testing situation a bit even if not quite all the way.

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h]: FlutterViewController_Internal.h

Comment (justinmc) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h]: Ah thanks, I missed that.

Though it seems to also give me errors when the test tries to mock it.  I just copied this exactly over to the _Internal header.  Is there something else I need to do to make it accessible to the test?

```
FlutterViewControllerTest.mm:32:31: error: no known instance method for selector 'pasteboard'
```

Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h]: Did you include that header in the test file?

Comment (justinmc) [reply shell/platform/darwin/macos/framework/Headers/FlutterViewController.h]: That was it, thanks.

**Comment by justinmc** [comment]
Waiting on https://github.com/flutter/engine/pull/21111 to fix the Mac iOS Engine test.

**Comment by justinmc** [comment]
Finally green 🎉 

@stuartmorgan Ready for final review when we're back on Monday.

**Comment by justinmc** [comment]
@stuartmorgan Reminder that this needs a final review when you have a chance.

**Review by stuartmorgan — APPROVED** [review]

**Pull request merged**
